### ISSUE TITLE
RMSNorm: Inconsistent output w/ and w/o JIT

### ISSUE THREAD

**Issue by ridcl** [google-deepmind/gemma#367]
I'm working on [NNX version](https://github.com/ridcl/fabrique/blob/gemma/src/fabrique/models/gemma/) of Gemma and noticed inconsistency between the two implementations of `RMSNorm`. After digging a bit, I found out that there's also slight inconsistency in Linen implementation with and without JIT compilation when used with `bfloat16`. Reproducible example below:

```python
import jax
import jax.numpy as jnp
import flax.linen as nn
from gemma.gm.nn._layers import RMSNorm
from gemma.gm.utils import _dtype_params


batch_size = 2
num_featues = 32
dtype = jnp.bfloat16
key_x, key_norm = jax.random.split(jax.random.key(104))
x = jax.random.normal(key_x, (batch_size, num_featues), dtype=jnp.bfloat16)

with _dtype_params.initialize_param_with_dtype(dtype):
  norm = RMSNorm()
  # norm.init() would set 'scale' to all zeros; we initialize it to random instead
  variables = {"params": {"scale": jax.random.normal(key_norm, (num_featues,), dtype=dtype)}}

  out_jit = jax.jit(norm.apply)(variables, x)
  out = norm.apply(variables, x)
  assert (out_jit == out).all()             # assertion fails
```
Some observations:

* the max difference between jitted and non-jitted output is `0.015625`
* result equivalence:
   - `nnx_norm(x) == linen_norm.apply(..., x)`
   - `nnx_norm(x) != nnx.jit(nnx_norm)(x)`
   - `linen_norm.apply(..., x) != jax.jit(linen_norm.apply)(..., x)`
   - `nnx.jit(nnx_norm)(x) != jax.jit(linen_norm.apply)(..., x)`

Questions:
* Is this behavior expected? I understand that operations in `bfloat16` are subject to slight discrepancies, but it's strange that it doesn't happen to other operations and to the NNX version. 
* Is `bfloat16` even correct dtype for norm? I base my analysis on usage of `_dtype_params.initialize_param_with_dtype` in [`_transformer.py`](https://github.com/google-deepmind/gemma/blob/main/gemma/gm/nn/_transformer.py#L226-L236), but it's pretty much possible I just haven't understood the control flow and `RMSNorm` is supposed to be in `float32`.
* If the answers to both questions above are "yes", do you have any suggestion on how to get consistent output between Linen and NNX versions?

**Comment by ridcl** [comment]
After more experiments I see that equality of NNX results w/ and w/o JIT actually depends on the random seed. 

<details>
  <summary>Full code of Linen and NNX implementations and a few tests</summary>

```python
import jax
import jax.numpy as jnp
from flax import linen as nn
from flax import nnx
from gemma.gm.utils import _dtype_params


class GemmaRMSNorm(nnx.Module):
    """
    Gemma RMSNorm layer.

    This layer (Linen version) is used in the original Gemma implementation [1]
    and behaves slightly differently than nnx.RMSNorm. We use it here
    for compatibility with Gemma's weights.
    """

    def __init__(
        self,
        num_features: int,
        *,
        param_dtype: jax.typing.DTypeLike = jnp.bfloat16,
        rngs: nnx.Rngs | None = None,  # not used
    ):
        self.scale = nnx.Param(jnp.zeros(shape=num_features, dtype=param_dtype))

    def __call__(self, x):
        scale = self.scale
        var = jnp.mean(jnp.square(x), axis=-1, keepdims=True)

        # Jax.lax.rsqrt is used because it returns different floats than
        # jnp.reciprocal(jnp.sqrt(var + 1e-06))
        normed_inputs = x * jax.lax.rsqrt(var + 1e-06)

        # normed_inputs is a rank-K tensor, K > 1 (K is typically 2 or 3). scale is
        # a rank-1 tensor. To avoid implicit rank-promotion, reshape scale to
        # a (1, ..., 1, D) tensor, so the rank of scale matches normed_inputs.
        scale = jnp.expand_dims(scale, axis=range(len(x.shape) - 1))
        normed_inputs = normed_inputs * (1 + scale)
        return normed_inputs



class RMSNorm(nn.Module):
  """RMSNorm layer."""

  @nn.compact
  def __call__(self, x):
      scale = self.param('scale', nn.initializers.zeros_init(), (x.shape[-1]))
      var = jnp.mean(jnp.square(x), axis=-1, keepdims=True)

      # Jax.lax.rsqrt is used because it returns different floats than
      # jnp.reciprocal(jnp.sqrt(var + 1e-06))
      normed_inputs = x * jax.lax.rsqrt(var + 1e-06)

      # normed_inputs is a rank-K tensor, K > 1 (K is typically 2 or 3). scale is
      # a rank-1 tensor. To avoid implicit rank-promotion, reshape scale to
      # a (1, ..., 1, D) tensor, so the rank of scale matches normed_inputs.
      scale = jnp.expand_dims(scale, axis=range(len(x.shape) - 1))
      normed_inputs = normed_inputs * (1 + scale)
      return normed_inputs


def main():
    batch_size = 2
    num_featues = 32
    dtype = jnp.bfloat16
    seed = 104
    key_x, key_norm = jax.random.split(jax.random.key(seed))
    x = jax.random.normal(key_x, (batch_size, num_featues), dtype=jnp.bfloat16)

    with _dtype_params.initialize_param_with_dtype(dtype):
        norm_nn = RMSNorm()
        # norm.init() would set 'scale' to all zeros; we initialize it to random instead
        variables = {"params": {"scale": jax.random.normal(key_norm, (num_featues,), dtype=dtype)}}

        out_nn = norm_nn.apply(variables, x)
        out_nn_jit = jax.jit(norm_nn.apply)(variables, x)

    norm_nnx = GemmaRMSNorm(num_features=num_featues)
    norm_nnx.scale.value = variables["params"]["scale"]
    out_nnx = norm_nnx(x)
    out_nnx_jit = nnx.jit(norm_nnx)(x)
    out_nnx_jj = jax.jit(norm_nnx)(x)

    jnp.all(out_nn == out_nnx)
    jnp.all(out_nn == out_nn_jit)
    jnp.all(out_nnx == out_nnx_jit)
    jnp.all(out_nnx == out_nnx_jj)
    jnp.all(out_nn_jit == out_nnx_jit)
```

</details>
With `seed = 0`, I get:

```python
    jnp.all(out_nn == out_nnx)                 # True
    jnp.all(out_nn == out_nn_jit)              # False
    jnp.all(out_nnx == out_nnx_jit)          # True
    jnp.all(out_nnx == out_nnx_jj)            # True
    jnp.all(out_nn_jit == out_nnx_jit)       # False
```
with `seed = 104`:

```python
    jnp.all(out_nn == out_nnx)                 # True
    jnp.all(out_nn == out_nn_jit)              # False
    jnp.all(out_nnx == out_nnx_jit)          # False
    jnp.all(out_nnx == out_nnx_jj)            # False
    jnp.all(out_nn_jit == out_nnx_jit)       # False
```

So perhaps discrepancy between compiled and non-compiled versions is unavoidable. Yet, I'm still curious if there's a way to check consistency of Linen an NNX implementations.

**Comment by Balakrishna-Chennamsetti** [comment]
Hi @ridcl ,

The behavior you're seeing is a natural consequence of using JAX's JIT compilation and low-precision floating-point arithmetic. It's a powerful and efficient design choice, but it comes with a need to adjust your testing methods.

1. The discrepancies are most likely due to compiler optimizations and floating-point rounding, which are more pronounced in `bfloat16` data type.

2. Using `bfloat16` for RMSNorm is correct and intentional for the Gemma model.

3. The best way to check for consistency is to use `jnp.allclose` instead of `==` for comparisons and to compare the implementations in `float32` to isolate the `bfloat16` specific noise.

Thanks.

**Comment by ridcl** [comment]
Thanks for confirmation and clarification! 
The problem with `jnp.allclose()` is that it doesn't always help to find where discrepancies become more critical than just floating point rounding. For example, in my latest test of the whole transformer I compare output tokens, i.e. `output.logits.argmax(axis=-1)`, between NNX and Linen implementation. In `float32`, there's ~5% of different tokens, while in `bfloat16` the results differ in ~20% of cases. Neither of the results guarantees that the implementation is fully correct, but perhaps it's the best we can have for cross-framework model rewrites.
